File: dsa/Others/ransomNote.py
from collections import Counter

def can_construct(ransom_note: str, magazine: str) -> bool:
    # Count the frequency of each character in ransom_note and magazine
    ransom_note_count = Counter(ransom_note)
    magazine_count = Counter(magazine)

    # Debug: Print the frequency counts (similar to console.log in JS)
    print("Ransom Note Count:", ransom_note_count)
    print("Magazine Count:", magazine_count)

    # Check if each character in ransom_note is available in magazine with sufficient frequency
    for char, count in ransom_note_count.items():
        if magazine_count.get(char, 0) < count:
            return False  # If any character is insufficient, return False

    return True  # If all characters are available, return True

# Test the function
print(can_construct('bg', 'efjbdfbdgfjhhaiigfhbaejahgfbbgbjagbddfgdiaigdadhcfcj'))

File: dsa/Others/ransomNote.js
/**
 * @param {string} ransomNote
 * @param {string} magazine
 * @return {boolean}
 */
var canConstruct = function (ransomNote, magazine) {
  let ransomNoteObj = {};
  let magazineObj = {};
  let isAvail = false;

  for (let i = 0; i < ransomNote.length; i++) {
    if (ransomNoteObj[ransomNote[i]] === undefined) {
      ransomNoteObj[ransomNote[i]] = 1;
    } else {
      ransomNoteObj[ransomNote[i]]++;
    }
  }

  for (let j = 0; j < magazine.length; j++) {
    if (magazineObj[magazine[j]] === undefined) {
      magazineObj[magazine[j]] = 1;
    } else {
      magazineObj[magazine[j]]++;
    }
  }

  console.log(ransomNoteObj);
  console.log(magazineObj);
  for (let key in ransomNoteObj) {
    if (key in magazineObj) {
      if (ransomNoteObj[key] <= magazineObj[key]) {
        isAvail = true;
      } else {
        isAvail = false;
        break;
      }
    } else {
      isAvail = false;
      break;
    }
  }
  return isAvail;
};

console.log(
  canConstruct('bg', 'efjbdfbdgfjhhaiigfhbaejahgfbbgbjagbddfgdiaigdadhcfcj')
);
